equational_theories/EquationalResult.lean

```diff
@@ -0,0 +1,103 @@
+initialize equationalResultAttr : Unit ←
+  registerBuiltinAttribute {
+    name  := `equational_result
+    descr := equationalResultHelpString
+    applicationTime := AttributeApplicationTime.afterCompilation
+    add   := fun declName _stx _attrKind => do
+       let filename := (←read).fileName
+       discard <| Meta.MetaM.run do
+       let info ← getConstInfo declName
+       let entry ← match info with
+                   | .thmInfo  (val : TheoremVal) =>
+                     if let some imp ← parseImplication val.type then
+                       pure <| ⟨val.name, filename, .implication imp⟩
+                     else if let some nimp ← parseNonimplication val.type then
+                       pure <| ⟨val.name, filename, .nonimplication nimp⟩
+                     else if let some uncond ← parseUnconditionalEquation val.type then
+                       pure <| ⟨val.name, filename, .unconditional uncond⟩
+                     else
+                       throwError "failed to match type of @[equational_result] theorem"
+                   | _ => throwError "@[equational_result] is only allowed on theorems"
+       modifyEnv fun env =>
+         equationalResultsExtension.addEntry env entry
+    erase := fun _declName =>
+      throwError "can't remove `equational_result` attribute (not implemented yet)"
+  }
+
+/-- Returns the contents of the equational results environment extension.
+-/
+def extractEquationalResults {m : Type → Type} [Monad m] [MonadEnv m] [MonadError m] :
+    m (Array Entry) := do
+  return equationalResultsExtension.getState (← getEnv)
+
+/-- Prints the contents of the equational results environment extension.
+-/
+syntax (name := printEquationalResults) "#print_equational_results" : command
+
+elab_rules : command
+| `(command| #print_equational_results) => do
+  let rs ← extractEquationalResults
+  for ⟨name, _filename, res⟩ in rs do
+    match res with
+    | .implication ⟨lhs, rhs⟩ => println! "{name}: {lhs} → {rhs}"
+    | .nonimplication ⟨lhs, rhs⟩ => println! "{name}: ¬ ({lhs} → {rhs})"
+    | .unconditional rhs => println! "{name}: {rhs} holds unconitionally"
+
+--- Output of the extract_implications executable.
+structure Output where
+  implications : List Implication
+  nonimplications : List Implication
+  unconditionals : List String
+deriving Lean.ToJson, Lean.FromJson
+
+def collectResults {m : Type → Type} [Monad m] [MonadEnv m] [MonadError m] :
+    m Output := do
+  let rs := equationalResultsExtension.getState (← getEnv)
+  let mut implications : List Implication := []
+  let mut nonimplications : List Implication := []
+  let mut unconditionals : List String := []
+  for ⟨_name, _filename, res⟩ in rs do
+    match res with
+    | .implication imp => implications := imp::implications
+    | .nonimplication nimp => nonimplications := nimp::nonimplications
+    | .unconditional s => unconditionals := s::unconditionals
+  return ⟨implications, nonimplications, unconditionals⟩
```

Comment:
Better to use `Array` here, we expect these to be maybe not neglectible size.

Reply:
done in https://github.com/teorth/equational_theories/pull/93/commits/a4d113eef09a2db84f5bf19aee4fceced68e2d17